src/encrypted-body-middleware.js

```diff
@@ -0,0 +1,37 @@
+/**
+
+Copyright 2018 New Vector Ltd.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+**/
+
+const { decryptBody } = require('./decrypt-body.js');
+
+function encryptedBodyMiddleware(req, res, next) {
+    // If a POST body contains `encrypted_body`, decrypt it and pass it as a body
+    // to the next middleware
+
+    if (req.method === 'POST' && req.body && req.body.encrypted_body) {
+        try {
+            req.body = decryptBody(req.body.encrypted_body);
+        } catch (err) {
+            next(err);
+            return;
+        }
+    }
```

Comment:
Mmm, this reminds me a bit of a spring webapp where you'd stare at the thing for ages wondering why the request you're seeing doesn't match what the client is sending, and then realise it's being mangled by some bit of middleware installed somewhere the other side of the codebase. It's a cheap way of separating concerns without actually making dedicated interfaces, the drawback being you end up with fake HTTP request objects like this.

Meh, if this is the done thing in express then sobeit, but I do find it can get quite confusing when your apps get larger.

Reply:
It does feel quite bad to have 
```js
if (req.method === 'POST' && req.body && req.body.encrypted_body) 
   do this specific thing transparently
```

I don't think it is the done thing necessarily. Would it help at all if there was sufficient logging to make the manoeuvre less transparent (for the server-side debug)?

Reply:
Yeah, this might help actually.